### ISSUE TITLE
C# scanner does not check for missing exponent digits for real literals

### ISSUE THREAD

**Issue by leppie** [dotnet/roslyn#6817]
When I attempted to fix #6109, a look in the code showed that the scanner does not check for real literals correctly.

More specifically, it allows numbers like the following to pass to subsequent stages of the compiler:
* `0e`
* `.5e`
* `1e+`

The bug is here: https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Parser/Lexer.cs#L1032

While the bug is easy to fix, the bigger problem is what error to use in the case. In the past `CS0594` has been used, but this is ambiguous and misleading. There seems to be no general error for invalid reals as `CS1013` for integers. Perhaps one should be created. Another option would be to extend `CS0594` to say `Invalid real literal or [original error message]`.

Sidenote: The legacy compiler actually accepts `0e` as valid.

**Comment by gafter** [comment]
Agree. This issue is one of code cleanliness (the compiler should catch errors earlier in the pipeline rather than later) and unclear error messages.

**Comment by mburbea** [comment]
Amazingly, these syntax parse errors in C# actually are valid and understood by sql server, and are valid ways to declare `float` literals in t-sql.
```sql
select  0e, 
           .5e, 
           1e+,
           1e-
```

**Comment by gafter** [comment]
Closed in https://github.com/dotnet/roslyn/commit/27bd018e4a515c3672d89dc9afa8b2426e4df148
